Add ECR credential provider

therc opened the pull request:
This addresses #19113 (how to use ECR). It requires and for the time being incorporates #19351 (AWS SDK 1.0.7)
Open items:
- [x] how to interact with the AWS cloudprovider, if at all
  - [x] keep the duplicate awsHandleLogger?
- [x] figure what to do with the email address in the docker config
- [x] tests (coming)

cc: @justinsb

Commit:
Add ECR dependencies and refresh Godeps

Commit:
ECR credential provider

therc commented:
Forgot to mention that this requires the ecr:GetAuthorizationToken permission.

therc commented:
I added testing. Still not sure what to do with the email address and interaction with the AWS cloudprovider. Feedback welcome.

erictune commented:
LGTM

erictune commented:
@k8s-bot add to whitelist

therc commented:
I like that. It looks like Docker will get rid of the email address, which was used to automatically create Hub accounts: https://github.com/docker/docker/issues/18668

erictune commented:
What you did with the email address seems fine.  It is similar to what happens on GCE/GCR.

erictune commented:
Can you add a few sentences to `docs/user-guide/images.md`.  Maybe put it after `### Using Google Container Registry`.

Maybe say something like this (please review for correctness):
```
### Using AWS ECR
Kubernetes has native support for the [AWS ECR](https://aws.amazon.com/ecr/) when nodes are AWS instances.

Simply use the full image name (e.g. $FIXME/imagename:tag) in the Pod definition.

All users of the cluster who can create pods will be able to run pods that use any
of the images in the ECR registry.

The kubelet will fetch and periodically refresh GCR credentials.  It needs the  `ecr:GetAuthorizationToken` to do this.

```

therc commented:
https://github.com/kubernetes/kubernetes/blob/7a8e7d01115c2388ba560d959c26c59db54be2b6/docs/user-guide/images.md

erictune commented:
I don't have a problem with you creating a cloud provider here to get the zone and region.

erictune commented:
LGTM.  You may want to run "make verify".  I suspect this will tell you that you need to run hack/update-generated-docs.sh to regenerated the TOC of the doc you edited.

erictune commented:
Thanks for this contribution!

erictune commented:
Yep.  Needs a run of hack/update-generated-docs.sh

therc commented:
I fixed the TOC, added the read-only ECR permissions to the node policy, so that it can work out of the box, and added a few comments about that to `aws_under_the_hood.md`. Making sure @justinsb sees the policy change so it's not a surprise.

eliaslevy commented:
Are these privileged really needed?
```
"ecr:BatchCheckLayerAvailability",
"ecr:GetDownloadUrlForLayer",
"ecr:GetRepositoryPolicy",
"ecr:DescribeRepositories",
"ecr:ListImages",
"ecr:BatchGetImage"
```

My understanding is that those are needed if you are using the AWS APIs to fetch images.  But if you are using docker pull you should only need `ecr:GetAuthorizationToken`, no?

therc commented:
@eliaslevy I remember trying with just that one permission initially and it didn't work. There's probably a subset that will do, but I didn't do the bisection work to find it. From the sound of it, I suspect that at least the ones with Layer in it are needed, but that's just a wild guess.

erictune commented:
Let's get this in and then I'll welcome follow-up PRs that reduce the permission set.

erictune commented:
LGTM.

erictune commented:
The presubmit checker says `Your godeps changes are not reproducable`.

Can you try running `hack/verify-godeps.sh` and try to fix whatever it says?

Here is the detailed message:

```
hack/verify-godeps.sh master
Starting to download all kubernetes godeps. This takes a while
Download finished
Your godeps changes are not reproducable
diff -Naupr '--exclude=symlink' /home/travis/gopath/src/github.com/kubernetes/kubernetes/Godeps/_workspace/src/github.com/stretchr/testify/LICENCE.txt /tmp/gopath.U8INxX/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/stretchr/testify/LICENCE.txt
--- /home/travis/gopath/src/github.com/kubernetes/kubernetes/Godeps/_workspace/src/github.com/stretchr/testify/LICENCE.txt	1970-01-01 00:00:00.000000000 +0000
+++ /tmp/gopath.U8INxX/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/stretchr/testify/LICENCE.txt	2016-01-22 17:55:38.479121590 +0000
@@ -0,0 +1,9 @@
+Copyright (c) 2012 - 2013 Mat Ryer and Tyler Bunnell
+
+Please consider promoting this project if you find it useful.
+
+Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
\ No newline at end of file
Removing /tmp/gopath.U8INxX
make: *** [verify] Error 1
```

I think maybe it is saying something about a missing newline in the LICENSE file or something.
It's possible this is a bug in our script.

erictune commented:
Thats wierd, though, because you didn't touch the file it is complaining about.

therc commented:
That's what I was wondering..

erictune commented:
@k8s-bot please travis build this again

erictune commented:
@therc did you just rebase?  Either you rebased wrong or github is confused.  It thinks there are 23 commits.

erictune commented:
If you `git rebase -i HEAD~2` and then reword one of your commits and then push again, sometimes it fixes things if github is confused.

therc commented:
Yeah, wrong rebase. verify-godeps wanted an 'origin' remote, but I had kubernetes/kubernetes as 'upstream'. Fixed.

therc commented:
I get the same error. I ran godep again, which picked up a new go-dockerclient. Running verify-godeps.sh again. Still making no sense to me.

therc commented:
Still the same error. The file is missing in master, too. Not sure what to do; it looks like something flakey to me.

erictune commented:
@mikedanese explained that the godeps error is probably from a previous submit.  The godeps are only verified on submits that modify godeps. Not your fault.  Merging and letting someone who is a godeps expert sort it out.

erictune commented:
thanks for the contribution!

mgoodness commented:
Yes, thank you very much, @therc! My company just started using ECR, and I hope to get K8s in the door soon. This will go a long way towards making that happen.

caesarxuchao commented:
I think this breaks hack/test-cmd.sh when it's running locally on Linux. Not sure why Jenkins tests pass.

therc commented:
How does it break? Does it happen when there are local AWS credentials?

caesarxuchao commented:
When starting kubelet in hack/test-cmd.sh, `newAWSCloud()` is triggered after this PR got merged, and emits `Zone not specified in configuration file; querying AWS metadata service`, then the kubelet never succussfully starts.

full log of running `hack/test-cmd.sh` is here: http://pastebin.com/wcZbGdVG

caesarxuchao commented:
> Does it happen when there are local AWS credentials?

I don't have that on my machine.

erictune commented:
Looks like  NewDockerKeyring() in pkg//credentialprovider/plugins.go calls
Enabled, and aws impl of Enabled() tries to connect to AWS.

Enabled() needs to decide if it should connect to AWS or not.


On Mon, Jan 25, 2016 at 4:01 PM, Chao Xu <notifications@github.com> wrote:

> Does it happen when there are local AWS credentials?
>
> I don't have that on my machine.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/19447#issuecomment-174736638>
> .
>

therc commented:
I think I misinterpreted how GetCloudProvider works, then. I am not sure how this would be handled, other than exposing the --cloud-provider flag globally. Or perhaps renaming aws_credentials::init to something else and getting it called from the AWS cloudprovider.

erictune commented:
Actually, I am not sure this PR is at fault.  Most of our tests are green,
and this hasn't been reverted.  Why would it cause intermittent failure.
Maybe that just happened to be the last log message before timeout?

On Mon, Jan 25, 2016 at 4:11 PM, Rudi C <notifications@github.com> wrote:

> I think I misinterpreted how GetCloudProvider works, then. I am not sure
> how this would be handled, other than exposing the --cloud-provider flag
> globally. Or perhaps renaming aws_credentials::init to something else and
> getting it called from the AWS cloudprovider.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/19447#issuecomment-174739804>
> .
>

erictune commented:
Maybe the kubelet does eventually come up, but it takes longer after this
change?

On Mon, Jan 25, 2016 at 4:13 PM, Eric Tune <etune@google.com> wrote:

> Actually, I am not sure this PR is at fault.  Most of our tests are green,
> and this hasn't been reverted.  Why would it cause intermittent failure.
> Maybe that just happened to be the last log message before timeout?
>
> On Mon, Jan 25, 2016 at 4:11 PM, Rudi C <notifications@github.com> wrote:
>
>> I think I misinterpreted how GetCloudProvider works, then. I am not sure
>> how this would be handled, other than exposing the --cloud-provider flag
>> globally. Or perhaps renaming aws_credentials::init to something else and
>> getting it called from the AWS cloudprovider.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/19447#issuecomment-174739804>
>> .
>>
>
>

therc commented:
My Enabled() code assumes that GetCloudProvider will return something only if --cloud-provider=aws, but I think that it just ends up forcibly calling newAWSCloud(), no matter what.

caesarxuchao commented:
I tested hack/test-cmd.sh before and after the merging commit of this PR, and it makes the difference.
I don't understand it's not an issue on Jenkins.

> Maybe the kubelet does eventually come up, but it takes longer after this
change?
Could be, hack/test-cmd.sh fails after a timeout.

therc commented:
Does #20129 fix it?

erictune commented:
Maybe because both GCE and AWS VMs have http://169.254.169.254/ and so the
AWS library doesn't have to wait for a response from that URL.  but on
desktop it times out.

On Mon, Jan 25, 2016 at 4:26 PM, Rudi C <notifications@github.com> wrote:

> Does #20129 <https://github.com/kubernetes/kubernetes/pull/20129> fix it?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/19447#issuecomment-174745156>
> .
>

erictune commented:
That fix looks fine.  Just asked for a comment.

On Mon, Jan 25, 2016 at 4:29 PM, Eric Tune <etune@google.com> wrote:

> Maybe because both GCE and AWS VMs have http://169.254.169.254/ and so
> the AWS library doesn't have to wait for a response from that URL.  but on
> desktop it times out.
>
> On Mon, Jan 25, 2016 at 4:26 PM, Rudi C <notifications@github.com> wrote:
>
>> Does #20129 <https://github.com/kubernetes/kubernetes/pull/20129> fix it?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/19447#issuecomment-174745156>
>> .
>>
>
>

caesarxuchao commented:
Thanks for the quick fix!

aveshagarwal commented:
I also see that there hack/test-cmd.sh failing due to this:

```
I0126 12:22:32.854106    6873 aws.go:466] Zone not specified in configuration file; querying AWS metadata service
^C+++ [0126 12:22:35] Clean up complete
!!! Error in /root/kubernetes/hack/lib/util.sh:45
  'sleep ${wait}' exited with status 130
Call stack:
  1: /root/kubernetes/hack/lib/util.sh:45 kube::util::wait_for_url(...)
  2: ./hack/test-cmd.sh:144 main(...)
Exiting with status 1
```

innovia commented:
@therc thanks for this PR, how can i get it on my running ec2 instances, i looked for the files but can't find them

therc commented:
@innovia if you mean a new kubelet binary, which is all that should be needed other than an instance policy update, I think you'll want the next alpha. They seem to be released every few weeks. I'm not sure what's the actual release schedule for alphas, if any.

innovia commented:
@therc if the only thing i need for now is just kubelet binary on the minions  - how do i get / compile juts that binary?

therc commented:
@innovia: try

`make all WHAT=cmd/kubelet`

I'm not sure about compatibility between a much newer node against an older master, perhaps it's safer if you apply the change to a source tree as close as possible to what you're already running.

kzwang commented:
@therc Seems this is not working if the cluster is not in `us-east-1` region even if the image is using `us-east-1` ECR. From the code seems it's using the cluster's region for get the token, is it possible to read the ECR region from the image URL?

justinsb commented:
@kzwang is this the same problem as #21783?  If not, can you please open a new issue and tag myself & therc.

21783 is "just" a delay on first startup, AFAICT.

innovia commented:
@therc do you think there's a way to run locally on virtualbox image, I've setup the env and .aws/credentials but its failing to authenticate with ECR from local VM

```bash
docker run \
    --volume=/:/rootfs:ro \
    --volume=/sys:/sys:ro \
    --volume=/var/lib/docker/:/var/lib/docker:rw \
    --volume=/var/lib/kubelet/:/var/lib/kubelet:rw \
    --volume=/var/run:/var/run:rw \
    --volume=/home/ubuntu/.aws:/root/.aws \
    --net=host \
    --pid=host \
    --privileged=true \
    --env AWS_ACCESS_KEY_ID=myAwsAccessKey \
    --env AWS_SECRET_ACCESS_KEY="myAwsSecretKey" \
    -d \
    gcr.io/google_containers/hyperkube-amd64:v1.2.0-beta.1 \
    /hyperkube kubelet \
        --containerized \
        --hostname-override="127.0.0.1" \
        --address="0.0.0.0" \
        --api-servers=http://localhost:8080 \
        --config=/etc/kubernetes/manifests \
        --cluster-dns=10.0.0.10 \
        --cluster-domain=cluster.local \
        --allow-privileged=true --v=10 \
        --cloud-provider=aws       
```
